pkg/pool-weighted/contracts/smart/ManagedPool.sol

```diff
@@ -674,6 +741,71 @@ contract ManagedPool is BaseWeightedPool, ProtocolFeeCache, ReentrancyGuard {
         _require(tokenData != 0, Errors.INVALID_TOKEN);
     }
 
+    // Join/exit callbacks
+
+    function _beforeJoinExit(
+        uint256[] memory,
+        uint256[] memory,
+        uint256
+    ) internal virtual override {
+        // The AUM fee calculation is based on inflating the Pool's BPT supply by a target rate.
+        // We then must collect AUM fees whenever joining or exiting the pool to ensure that LPs only pay AUM fees
+        // for the period in which they are an LP within the pool, otherwise an LP could shift their share of AUM fees
+        // onto the remaining LPs in the pool by exiting before it was paid.
+        _collectAumManagementFees();
+    }
+
+    /**
+     * @dev Calculates the AUM fees accrued since the last collection and pays it to the pool manager.
+     * This function is called automatically on joins and exits.
+     */
+    function _collectAumManagementFees() internal returns (uint256 bptAmount) {
+        uint256 lastCollection = _lastAumFeeCollectionTimestamp;
+        uint256 currentTime = block.timestamp;
+
+        // Collect fees based on the time elapsed
+        if (currentTime > lastCollection) {
+            // Reset the collection timer to the current block
+            _lastAumFeeCollectionTimestamp = currentTime;
+
+            uint256 managementAumFeePercentage = getManagementAumFeePercentage();
+
+            // If `lastCollection` has not been set then we don't know what period over which to collect fees.
+            // We then perform an early return after initializing it so that we can collect fees next time.
+            // We also perform an early return if the Pool has been paused or the or the AUM fee is zero.
```

Comment:
```suggestion
            // If `lastCollection` has not been set then we don't know what period over which to collect fees.
            // We then perform an early return after initializing it so that we can collect fees next time. This
            // means that AUM fees are not collected for any tokens the Pool is initialized with until the first
            // non-initialization join or exit.
            // We also perform an early return if the AUM fee is zero, to save gas.
            //
            // If the Pool has been paused, all fee calculation and minting is skipped to reduce execution
            // complexity to a minimum (and therefore likelihood of errors). We do still update the last
            // collection timestamp however, to avoid potentially collecting extra fees if the Pool were to
            // be later unpaused. Any fees that would be collected while the Pool is paused are lost.
```